### ISSUE TITLE
Failing Validation: Inheritance

### ISSUE THREAD

**Issue by pk-picturepark** [NJsonSchema/NJsonSchema#224]
There seems to be a problem with the validation of inherited values. 

By adding the following two lines of code 

```
var schema = JsonSchema4.FromType<Container>();
var errors = schema.Validate(json);
```

at the end of the test method _When_JsonInheritanceConverter_is_used_then_inheritance_is_correctly_serialized_and_deserialized()_ validation will return 1 error. 

```
{NotOneOf: #/Animal
{
  NoAdditionalPropertiesAllowed: #/Animal.Bar
  NoAdditionalPropertiesAllowed: #/Animal.SubElements
}
{
  NullExpected: #/Animal
}
}
```

Am I doing something wrong or is this a known issue?

**Comment by rsuter** [comment]
Yes, this is a problem.. but because I mainly use NJsonSchema for code generation (http://nswag.org) it was not an issue...

**Comment by rsuter** [comment]
The problem is that for validation we need oneOf inheritance (ie define the child classes): https://github.com/NJsonSchema/NJsonSchema/issues/13

... and for swagger generation we need allOf inheritance (ie define parent classes)

**Comment by rsuter** [comment]
But oneOf inheritance is not supported by swagger (oneOf is not defined)

**Comment by pk-picturepark** [comment]
As there is yet no real property inheritance / overwrite support in the json schema standard I handle validation with a custom json schema structure. 

For instance: I have two classes _Person_ and _Player_. The _Player_ class inherits all the properties from the _Person_ class. Validaton should pass for instances of both classes. AddtionalProperties should be restricted.

My working json schema looks as follows:

`{
	"additionalProperties": true,
	"definitions": {
		"Person": {
			"additionalProperties": false,
			"properties": {
				"PersonProperty": {
					"type": "string"
				},
				"Type": {
					"type": "string",
					"enum": ["Person"]
				}
			}
		},
		"Player": {
			"additionalProperties": false,
			"properties": {
				"PersonProperty": {
					"type": "string"
				},
				"PlayerProperty": {
					"type": "string"
				},
				"Type": {
					"type": "string",
					"enum": ["Player"]
				}
			}
		}
	},
	"oneOf": [{
		"$ref": "#/definitions/Person"
	},
	{
		"$ref": "#/definitions/Player"
	}]
}`

As you can see, I have to duplicate the _Person_ properties within the _Player_ schema and use a enum restricted _Type_ string property as suggested [here](http://json-schema.org/example2.html). 

Leaving out the _allOf_ reference in the _Player_ schema to the _Person_ schema will somehow break the inheritance configuration, but using allOf does then only pass validation if I set _addtionalProperties_ to true on the _Person_ and _Player_ schemas. Something I want to avoid.

I also tried out other _allOf_, _oneOf_ configurations and relocations of the _Type_ property without any success. Property inheritance and overwrites are not yet supported.

There might be a better approach here and I would be happy for any inputs. However, I am sure that the output provided by NJsonSchema does not yet create the json schema structure for the validations I need.

**Comment by rsuter** [comment]
Did you also try the FlattenInheritanceHierarchy option?

**Comment by rsuter** [comment]
For validation we need: https://github.com/NJsonSchema/NJsonSchema/issues/13
